### ISSUE TITLE
v. 0.7 - random parsing 

### ISSUE THREAD

**Issue by abialy** [lark-parser/lark#343]
Hi, 
I found an issue with parsing a specific grammar. In 0.6.5 the parsing is consistent meaning every time the same tree is produced. In 0.7 the tree is different each time it runs. 

The grammar can be found here:  https://pastebin.com/AHbwbcUJ
The script -> https://pastebin.com/ZUswjpMh

I run that test program: 

from lark import Lark, Transformer, Tree
from lark.tree import pydot__tree_to_png

parser = Lark(open('test.lark'), debug=True)

f = open('test.sst')
script = f.read()
tree = parser.parse(script)
print(tree.pretty())
pydot__tree_to_png(tree, "./tree.png")

Any idea why it happens? 

Regards,
A.

**Comment by jrouquie** [comment]
Maybe related to https://github.com/lark-parser/lark/issues/278

**Comment by pacoispaco** [comment]
I believe I have the same problem. That is, for a given grammar and input, I get one of two different parse trees each time I parse the input. It seems to be non-deterministic which parse tree is constructed. I also get the problem irrespective of which parser, 'early', 'lalr' or 'cyk', I choose.

I'm using 0.7.1 and had the same problem in 0.7.0.

I have managed to reduce my grammar, to a minimal grammar that still reproduces the problem: https://pastebin.com/hKKvfs9r

The input is:
```
not true or not false
```

The two different parse trees produced are as follows. This is the expected parse tree I expect, or want, to get:
```
expression
  normal_expression
    composite_expression
      expression
        negated_expression
          not
          expression
            normal_expression
              proposition	true
      connective	or
      expression
        negated_expression
          not
          expression
            normal_expression
              proposition	false
```

and this is the other parse tree, that I also get:

```
expression
  negated_expression
    not
    expression
      normal_expression
        composite_expression
          expression
            normal_expression
              proposition	true
          connective	or
          expression
            negated_expression
              not
              expression
                normal_expression
                  proposition	false
```

**Comment by erezsh** [comment]
@pacoispaco That is very unlikely, the lalr implementation should be completely deterministic. If you want me to look into it, please provide a fully working example (inc. code, not just the grammar) that demonstrates your claim.

**Comment by pacoispaco** [comment]
@erezsh You are correct. I thought I had tested those parsers as well, but I obviously hadn't. LALR and CYK are deterministic. The problem with the Early-parser is real though.

I have witten a small CLI-program that reads LARK files and some language input and print the resulting parse tree to stdout. You can find that program here: https://pastebin.com/MLyzzjff 

The grammar I'm using and the language input that causes non-deterministic behaviour in the Earley-parser you can see in comment above. Simply run the program above with the grammar above, the starting rule 'expression' and the input 'not true or not false':
```
$ ./larkparse.py -eproblem-ebnf.lark -sexpression -pearley -l'not true or not false'
```
a few times and you will se different parse trees.

I will continue looking into the problem myself. My full parser still has some other ambiguities in it which I want to fix first.

**Comment by pacoispaco** [comment]
I have rewritten my grammar so it now can parse my grammar correctly. And it works with with all Lark parsers; 'early', 'lalr' or 'cyk'. 'lalr' is definitely the fastest parser for my grammar.

It would still be nice to know what causes the non-deterministic behaviour of the Early parser.

**Comment by erezsh** [comment]
LALR will always be the fastest. Non-determinism is an onging problem with Earley. I would love to have the time to solve it, but that might take a few days of work if not weeks, and I'm afraid I don't have the time right now. Perhaps one day!

**Comment by pacoispaco** [comment]
That is perfectly ok with me. It was just confusing for me initially, not having done any DSL/compiler/parsing stuff for quite some time. Sorry for the hiccup with my incorrect error reporting regarding the 'lalr' parser. And thanks again for a very nice and elegant tool!

**Comment by night199uk** [comment]
The problem here is that your grammar is fundamentally ambiguous. There is a bug in that it should be deterministic, but ideally you'd factor out your ambiguities. As you see from the two resulting parse trees, there are multiple ways to parse the expression. 

expression->normal_expression->composite_expression->expression->negated_expression can match 'not true'. Similarly, expression->negated_expression can also match not true.

You would want to rejig your grammar to remove the recursion here.

**Comment by jtrakk** [comment]
@erezsh  I often get random parsing when I'm using the Earley parser. Would it be helpful for your debugging to have a collection of examples? I can try to post them when they come up if you like.

**Comment by erezsh** [comment]
@jtrakk Yes, a collection of examples could be helpful. Especially if they are simplified to having few elements.

**Comment by night199uk** [comment]
Should also be fixed by #428.

**Comment by erezsh** [comment]
Presumably fixed! If this issue persists, feel free to re-open.
